### PR TITLE
implement get and describe cmd

### PR DESCRIPTION
**Pull request by Nexucis**
This PR is implementing the command `get` which will be used to list the resources from the API.

For example:

```bash
./bin/percli get dashboards
       NAME       | PROJECT |   AGE
------------------+---------+----------
  NodeExporter    | perses  | 106751d
  SimpleLineChart | perses  | 106751d

```

Or 

```bash
./bin/percli get gdts
          NAME          | DATASOURCE TYPE |   AGE
------------------------+-----------------+----------
  PrometheusDemo        | Prometheus      | 106751d
  PrometheusDemoBrowser | Prometheus      | 106751d
  PrometheusLocal       | Prometheus      | 106751d
```

Signed-off-by: Augustin Husson <husson.augustin@gmail.com>

### CODE REVIEW

**Comment by Nexucis** [comment]
I just add another command : `describe` which I think make sense to have it in the same PR.

The goal of the command `describe` is to show details about a specific resource. For example: 

```bash
./bin/percli describe dashboards NodeExporter -ojson
{"kind":"Dashboard","metadata":{"name":"NodeExporter","created_at":"0001-01-01T00:00:00Z","updated_at":"0001-01-01T00:00:00Z","project":"perses"},"spec":{"datasource":{"name":"PrometheusDemo","kind":"Prometheus","global":true},"duration":"6h","variables":{"job":{"kind":"LabelValuesQuery","displayed_name":"job","hide":false,"parameter":{"label_name":"job","matchers":["node_uname_info"],"capturing_regexp":"(.*)"}},"node":{"kind":"LabelValuesQuery","displayed_name":"node","hide":false,"parameter":{"label_name":"instance","matchers":["node_uname_info{job=\"$job\"}"],"capturing_regexp":"(.*)"}}},"panels":{"CPUBusy":{"displayed_name":"CPU Busy","kind":"GaugeChart","chart":{"expr":"(((count(count(node_cpu_seconds_total{instance=\"$node\",job=\"$job\"}) by (cpu))) - avg(sum by (mode)(rate(node_cpu_seconds_total{mode='idle',instance=\"$node\",job=\"$job\"}[5m])))) * 100) / count(count(node_cpu_seconds_total{instance=\"$node\",job=\"$job\"}) by (cpu))"}},"RamUsed":{"displayed_name":"Ram Used","kind":"GaugeChart","chart":{"expr":"100 - ((node_memory_MemAvailable_bytes{instance=\"$node\",job=\"$job\"} * 100) / node_memory_MemTotal_bytes{instance=\"$node\",job=\"$job\"})"}},"RootFSUsed":{"displayed_name":"Root FS Used","kind":"GaugeChart","chart":{"expr":"100 - ((node_filesystem_avail_bytes{instance=\"$node\",job=\"$job\",mountpoint=\"/\",fstype!=\"rootfs\"} * 100) / node_filesystem_size_bytes{instance=\"$node\",job=\"$job\",mountpoint=\"/\",fstype!=\"rootfs\"})"}},"SwapUsed":{"displayed_name":"Swap Used","kind":"GaugeChart","chart":{"expr":"((node_memory_SwapTotal_bytes{instance=\"$node\",job=\"$job\"} - node_memory_SwapFree_bytes{instance=\"$node\",job=\"$job\"}) / (node_memory_SwapTotal_bytes{instance=\"$node\",job=\"$job\"} )) * 100"}},"SysLoad15mAVG":{"displayed_name":"Sys Load (15m avg)","kind":"GaugeChart","chart":{"expr":"avg(node_load15{instance=\"$node\",job=\"$job\"}) /  count(count(node_cpu_seconds_total{instance=\"$node\",job=\"$job\"}) by (cpu)) * 100"}},"SysLoad5mAVG":{"displayed_name":"Sys Load (5m avg)","kind":"GaugeChart","chart":{"expr":"avg(node_load5{instance=\"$node\",job=\"$job\"}) /  count(count(node_cpu_seconds_total{instance=\"$node\",job=\"$job\"}) by (cpu)) * 100"}}},"layouts":{"main":{"kind":"Expand","parameter":{"open":true,"children":[{"$ref":"#/spec/panels/CPUBusy"},{"$ref":"#/spec/panels/SysLoad5mAVG"},{"$ref":"#/spec/panels/SysLoad15mAVG"},{"$ref":"#/spec/panels/RamUsed"},{"$ref":"#/spec/panels/SwapUsed"},{"$ref":"#/spec/panels/RootFSUsed"}]}}},"entrypoint":{"$ref":"#/spec/layouts/main"}}}
```

**Review by sjcobb — APPROVED** [review]
I don't have any feedback on your go code :laughing: but this is really cool :rocket:

File: internal/cli/cmd/describe/describe.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+	name            string
+	output          string
+	resourceService cmdUtilsService.Service
+}
+
+func (o *option) complete(args []string) error {
+	if len(args) < 1 {
+		return fmt.Errorf(cmdUtils.FormatAvailableResourcesMessage())
+	} else if len(args) < 2 {
+		return fmt.Errorf("you have to specify the name of the resource you have to describe")
```
Comment (juliusv) [inline internal/cli/cmd/describe/describe.go:37]:
```suggestion
		return fmt.Errorf("please specify the name of the resource you want to describe")
```

Sounds friendlier? :)

File: internal/cli/cmd/describe/describe.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+	if err != nil {
+		return err
+	}
+	return cmdUtils.HandleOutput(o.output, entity)
+}
+
+func NewCMD() *cobra.Command {
+	o := &option{}
+	cmd := &cobra.Command{
+		Use:   "describe [SUBTYPE] [NAME]",
```
Comment (juliusv) [inline internal/cli/cmd/describe/describe.go:83]: How come "subtype" and not just "type" or "resource" type? Is there

File: internal/cli/cmd/describe/describe.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+		Short: "Show details of a specific resource",
+		Example: `
+## Describe a particular dashboards.
```
Comment (juliusv) [inline internal/cli/cmd/describe/describe.go:86]:
```suggestion
## Describe a particular dashboard.
```

File: internal/cli/cmd/describe/describe.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+percli describe dashboard nodeExporter
+
+## Describe a particular dashboards in a json.
```
Comment (juliusv) [inline internal/cli/cmd/describe/describe.go:89]:
```suggestion
## Describe a particular dashboard as a JSON object.
```

File: internal/cli/cmd/get/get.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+		return err
+	}
+	o.resourceService = svc
+	return nil
+}
+
+func (o *option) validate() error {
+	// check if project should be defined (through the config or through the flag) for the given resource.
+	if !o.allProject && len(o.project) == 0 && !cmdUtils.IsGlobalResource(o.kind) {
+		return fmt.Errorf("no project has been defined for the scope of this command. If you intended to get all resources across the different project, please use the flag --all")
```
Comment (juliusv) [inline internal/cli/cmd/get/get.go:73]:
```suggestion
		return fmt.Errorf("no project has been defined for the scope of this command. If you intended to get all resources across different projects, please use the flag --all")
```

File: internal/cli/cmd/get/get.go [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+# List all dashboards in the current project selected.
+percli get dashboards 
+
+# List all dashboards that begin with a given name in the current project selected.
+percli get dashboards node
+
+# List all dashboards in a specific project.
+percli get dashboards -p my_project
+
+#List all dashboards as a json object.
```
Comment (juliusv) [inline internal/cli/cmd/get/get.go:114]:
```suggestion
# List all dashboards as a JSON object.
```

File: internal/cli/cmd/get/get.go [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+percli get dashboards -a -ojson
+
+`,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdUtils.HandleError(o.complete(args))
+			cmdUtils.HandleError(o.validate())
+			cmdUtils.HandleError(o.execute())
+		},
+	}
+	cmd.Flags().BoolVarP(&o.allProject, "all", "a", o.allProject, "If present, list the request object(s) across all projects. Project in current context is ignored even if specified with --project.")
```
Comment (juliusv) [inline internal/cli/cmd/get/get.go:124]:
```suggestion
	cmd.Flags().BoolVarP(&o.allProject, "all", "a", o.allProject, "If present, list the requested object(s) across all projects. The project in the current context is ignored even if specified with --project.")
```

File: internal/cli/utils/resource.go [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+	return result, nil
+}
+
+// GetKind tries to find the kind from the given string. It returns an error if the kind is not managed,
+// or no alias exists.
+// This method is designed to be called in the cmds improving the readability of the kind arguments parsing.
+func GetKind(res string) (modelV1.Kind, error) {
+	alias := reverseResourceAliases()[strings.ToLower(res)]
+	if len(alias) <= 0 {
```
Comment (juliusv) [inline internal/cli/utils/resource.go:116]: Same comment about `= 0`.

File: internal/cli/utils/resource.go [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+	}
+	return result
+}
+
+// FormatAvailableResourcesMessage formats the available resources that the user can use
+func FormatAvailableResourcesMessage() string {
+	var result []string
+	for _, r := range resources {
+		var res string
+		if len(r.shortTerm) <= 0 {
```
Comment (juliusv) [inline internal/cli/utils/resource.go:144]: Dito

**Review by juliusv — COMMENTED** [review]
Looks good in general, just a couple of nits

File: internal/cli/cmd/describe/describe.go (hunk 2 above)
Comment (Nexucis) [reply internal/cli/cmd/describe/describe.go:83]: mmm that's an excellent question, I'm not sure either what I was thinking here. I will change it with `RESOURCE TYPE`

File: internal/cli/cmd/describe/describe.go (hunk 1 above)
Comment (Nexucis) [reply internal/cli/cmd/describe/describe.go]: Indeed, less threatening haha.

File: internal/cli/cmd/get/get.go [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+	allProject      bool
+	project         string
+	output          string
+	prefix          string
+	resourceService cmdUtilsService.Service
+}
+
+func (o *option) complete(args []string) error {
+	// first let's analyze the args to get what kind of resource we should get and if there is a prefix to use for the filtering.
+	if len(args) <= 0 {
```
[the remark this answers is not in the document]
Comment (Nexucis) [reply internal/cli/cmd/get/get.go]: ah yeah you are right. I think I did it like that (as a reflex) when thinking about what is it the opposite of `> 0`, well it's `<= 0`. That's like that I'm seeing it.
But yup you are right ofc, I fixed it.

**Comment by Nexucis** [comment]
thanks for your review @juliusv !

**Pull request merged**